rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,264 @@
+- only single-value scaling;
+- no run-time support.
+
+API:
+
+```c
+dnnl_status_t DNNL_API dnnl_post_ops_append_binary_v2(dnnl_post_ops_t post_ops,
+        float scale, dnnl_alg_kind_t alg_kind,
+        const dnnl_memory_desc_t *src1_desc);
+
+dnnl_status_t DNNL_API dnnl_post_ops_get_params_binary_v2(
+        const_dnnl_post_ops_t post_ops, int index, float *scale,
+        dnnl_alg_kind_t *alg_kind, const dnnl_memory_desc_t **src1_desc);
+```
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(scale, algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+```
+
+#### 2.2.2 Option 2
+
+An alternative solution would be to extend mechanism `attr::scales`. This
+solution doesn't require new API, but extends the existing one.
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+```
+
+While the alternative solution allows to avoid new API, it introduces one more
+way to set scale which complicates both implementation and integration.
+Please review section
+[Unifying scaling in long-term](#3-unifying-scaling-in-long-term) for more
+details.
+
+
+### 2.3. Scales for inputs in Concat
+
+The mechanism `attr:set_scales` will be introduced to Concat primitive. This
+mechanism is used in Binary primitive, so API is already there.
+
+The following limitations are in place for an initial implementation:
+ - no-runtime support;
+ - mask is 0.
+
+An example:
+
+```cpp
+    primitive_attr attr;
+    for (int n = 0; n < num_src; ++n)
+        attr.set_scales(DNNL_ARG_MULTIPLE_SRC + n, /* mask */ 0, src_scales[n]);
+```
+
+## 3. Unifying scaling in long-term
+
+The motivation behind unifying scaling:
+- Remove duplications;
+- Provide an aligned support across all use cases.
+
+There are a few issues with the current scaling support:
+ - Output scale, scales and binary_mul overlap each other;
+ - RNN quantization is specific to RNN, but can be replaced by other
+ scaling mechanisms:
+   ```
+   # Scaling semantics of set_scales and set_zero_points
+   x_f32 = scale * (x_i8 - zero_point)
+
+   # Scaling semantics of RNN quantization
+   x_i8 = scale * x_f32 + zero_point
+   x_f32 = 1 / scale * (x_i8 - zero_point)
+   ```
+   This will affect application side code as user should pass `1 / scale` to
+   oneDNN comparing to the current `scale`;
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+The changes:
+1. A unified scaling will be supported using mechanism `attr::set_scales` only:
+ - `attr::output_scales`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+ ```
+ - `attr::p_op::binary_mul`
+ ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
```

Comment:
Currently, for runtime scales/zp we require users to always pass a `memory` even if it's just a scalar. Since we plan to require users to always pass scales/zp at execution time then maybe it makes sense to try to simplify the API by allowing a simple scalar to be passed?

Reply:
To me it seems like not enough justification for changing oneDNN execution model. Memory based arguments are perfect for accelerators like GPU, because they have an engine and can incapsulate device-side memory objects like buffers. I am worried that adding a scalar argument might limit us in the future in case other accelerators will be added.

We can discuss keeping a primitive creation time single scale if there is a performance benefit.